src/librustc_typeck/collect.rs

```diff
@@ -1699,9 +1790,28 @@ fn find_opaque_ty_constraints(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
     }
 }
 
+/// Whether `ty` is a type with `_` placeholders that can be infered. Used in diagnostics only to
+/// use inference to provide suggestions for the appropriate type if possible.
+fn is_suggestable_infer_ty(ty: &hir::Ty) -> bool {
+    match &ty.kind {
+        hir::TyKind::Infer => true,
+        hir::TyKind::Slice(ty) | hir::TyKind::Array(ty, _) => is_suggestable_infer_ty(ty),
+        hir::TyKind::Tup(tys)
+            if !tys.is_empty()
+                && tys.iter().any(|ty| match ty.kind {
+                    hir::TyKind::Infer => true,
+                    _ => false,
+                }) =>
+        {
+            true
+        }
+        _ => false,
+    }
+}
+
 pub fn get_infer_ret_ty(output: &'_ hir::FunctionRetTy) -> Option<&hir::Ty> {
     if let hir::FunctionRetTy::Return(ref ty) = output {
-        if let hir::TyKind::Infer = ty.kind {
+        if is_suggestable_infer_ty(ty) {
```

Comment:
Are you sure `get_infer_ret_ty` called when we've already decided we're in an error branch?

Reply:
I was pretty sure, but I might be wrong. I'll double check.

Reply:
I'm pretty sure this is correct, it's only called from two places, below this and in `typeck_tables_of`(`_with_fallback`). I tried changing `typeck_tables_of` to not use it and inline that previous logic, but that leads to a cycle error when suggesting.